Separate usage of "b" and "c" fields

gzm0 opened the pull request:
"c" is now only used for Scala classes.
"b" is now only used for JS classes ("a" is the accessor).

Since we now generate ES6 classes in expression position, we adjust
ClosureAstTransformer to support this.

gzm0 commented:
@sjrd do we need to adjust `StackTrace.scala` and `StackTraceTest.scala` with this? (`StackTraceTest.scala` currently doesn't test JS classes, and I'd expect them to break with this, if they worked before).

sjrd commented on linker/shared/src/main/scala/org/scalajs/linker/backend/emitter/ClassEmitter.scala:
```diff
@@ -253,7 +241,7 @@ private[emitter] final class ClassEmitter(sjsGen: SJSGen) {
       val docComment = js.DocComment("@constructor")
       val ctorDef =
-        if (isJSClass) globalVarDef("c", className, ctorFun)
+        if (isJSClass) fileLevelVar("b", genName(className)) := ctorFun
```
Isn't `typeVar` still visible here?
```suggestion
        if (isJSClass) typeVar := ctorFun
```

sjrd commented on linker/shared/src/main/scala/org/scalajs/linker/backend/emitter/ClassEmitter.scala:
```diff
@@ -43,121 +43,113 @@ private[emitter] final class ClassEmitter(sjsGen: SJSGen) {
+
+      val entireClassDefWithGlobals = if (useClasses) {
+        genJSSuperCtor(tree).map { jsSuperClass =>
+          classValueVar := js.ClassDef(None, Some(jsSuperClass), allES6Defs)
```
We should still provide the `classValueIdent` as the `name` of the `js.ClassDef`, even when used in expression position. This makes sure that its `.name` property will be correctly filled in by the engine, and that overall the class and its methods appear in a better way in debugging tools.

sjrd commented:
> 
> 
> @sjrd do we need to adjust `StackTrace.scala` and `StackTraceTest.scala` with this? (`StackTraceTest.scala` currently doesn't test JS classes, and I'd expect them to break with this, if they worked before).

We should probably adjust `StackTrace.scala` for `$b_` prefixes, yes.

gzm0 commented on linker/shared/src/main/scala/org/scalajs/linker/backend/emitter/SJSGen.scala:
```diff
@@ -377,4 +377,18 @@ private[emitter] final class SJSGen(
     }
   }
+
+  def assignES5ClassMembers(classRef: Tree, members: List[MethodDef])(
```
It was there because of `genPropSelect`. I have move that one as wel.

gzm0 commented:
Updated. I have added a test in `StackTraceTest`. However, it doesn't fail after the change, because the frame gets detected on the static method (on the `s` field) that implements the exported method `"m"`.

IDK how to craft a test that doesn't fall victim to this.

sjrd commented:
Right. I can't imagine a way to craft such a test either. The good news is that it probably means that the bad behavior is indeed not observable anyway, and that therefore we don't need to change anything.

gzm0 commented:
Using `@JSName` might actually do the trick. We can then assert both names (the Scala name and the JS name). I'll try that.

gzm0 commented:
It worked :)
